### PR TITLE
Wave 3 complete

### PR DESCRIPTION
**Pull request by daphnegold**
Custom optionals coming soon!

### CODE REVIEW

File: lib/scrabble_master.rb [hunk 1]
```diff
@@ -1,2 +1,7 @@
 require_relative "scrabble"
 require_relative "player"
+require_relative "tilebag"
+require_relative "dictionary"
+require_relative "game"
+
+play_scrabble
```
Comment (Hamled) [inline lib/scrabble_master.rb]: It's not a good idea to place a method call to `play_scrabble` here, as every time a file `require`s anything from the Scrabble module, it will start playing the game.

For example, after the commit that introduced this (0c12df590a56e80650d6de3f81be8e318df4a424) it is not possible to run `rspec` on your project without starting the game (which doesn't seem to have an end?)

File: lib/tilebag.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
+module Scrabble
+  class TileBag
+
+    attr_accessor :tile_bag
+
+    def initialize
+      @tile_bag = (?A*9 + ?B*2 + ?C*2 + ?D*4 + ?E*12 + ?F*2 + ?G*3 + ?H*2 + ?I*9 + ?J +
+      ?K + ?L*4 + ?M*2 + ?N*6 + ?O*8 + ?P*2 + ?Q*1 + ?R*6 + ?S*4 + ?T*6 + ?U*4 +
+      ?V*2 + ?W*2 + ?X*1 + ?Y*2 + ?Z*1).split("").shuffle
```
Comment (Hamled) [inline lib/tilebag.rb:9]: As discussed in class this is probably fine for the Scrabble project and in future projects we'll have our static (and dynamic) data defined elsewhere outside of our code files.

However, in general it's a good (taste/style) rule to avoid using neat or quirky parts of a language just because they're available. The more time that another reader of your code must take to follow it, perhaps because you're using a language feature they've not encountered before, the more difficult it will be for them to understand your intention.

Also as a very, very minor side note: Due to the way that Ruby used to characters in 1.8 and prior versions, this code would not operate as intended on those versions. We will in the future see ways that you can build your projects to indicate a specific preferred Ruby version.

File: spec/player_spec.rb [hunk 3]
```diff
@@ -70,10 +78,30 @@
...
+      @tilebag = Scrabble::TileBag.new
+    end
+    it "draws 7 tiles" do
+      expect(@player.draw_tiles(@tilebag).length).to eq 7
+    end
+    it "adds 7 tiles to the players tiles array" do
+      @player.draw_tiles(@tilebag)
+      expect(@player.tiles.length).to eq 7
+    end
+    it "adds 3 tiles to the players tiles array already containing 4 tiles" do
```
Comment (Hamled) [inline spec/player_spec.rb:102]: A better description for this test would be something like "draws tiles from the tilebag until player has 7 tiles total".

File: spec/player_spec.rb [hunk 4]
```diff
@@ -70,10 +78,30 @@
...
+      @player.create_mock({tiles: %w(W O R D)})
+      @player.draw_tiles(@tilebag)
+      expect(@player.tiles.length).to eq 7
```
Comment (Hamled) [inline spec/player_spec.rb:105]: It might be a good idea to also expect `@tilebag.remaining_tiles` to have 3 fewer tiles than at the beginning of this test.

File: lib/player.rb [hunk 5]
```diff
@@ -1,17 +1,35 @@
...
+      word = word.upcase
       if won?
         return false
       else
-        @plays.push(word)
+        word_in_array = word.split("")
+        remaining_tiles = @tiles.dup
+
+        word_in_array.each do |l|
+          remaining_tiles.slice!(remaining_tiles.index(l)) if remaining_tiles.include?(l)
```
Comment (Hamled) [inline lib/player.rb:26]: Could you explain for me what this line of code is doing?

**Comment by Hamled** [comment]
Overall I like what y'all have accomplished with this project!

The test cases are all checking relevant core and edge cases, especially the section on the Player class for playing particular words.

However, I cannot run `rspec` on the latest version of this PR, so I am refraining from merging it at this time. Please feel free to fix the issue (for more information see my first line-specific comment) and add a comment here when RSpec tests are all passing.

File: lib/scrabble_master.rb (hunk 1 above)
Comment (daphnegold) [inline lib/scrabble_master.rb]: The game ends when you win at 100 points! 

This has been changed so you can run RSpec without starting the game. play_scrabble has been moved into game.rb, and require_relative "game" removed from scrabble_master.rb

File: lib/player.rb (hunk 5 above)
Comment (daphnegold) [inline lib/player.rb:26]: remaining_tiles is a local variable copy of @tiles, the tiles in the player's hand. The iterator goes through each letter in word_in_array (the word submitted by the user split into an array). The if statement checks if that letter even exists in the remaining_tiles array. This is necessary because otherwise index.("l") will return nil if the letter is not in remaining_tiles and we will get a TypeError. Once we know that 'l' is in remaining_tiles, we find its index (the first instance of it, if there are duplicates), then the 'slice!' method will remove the element at that index from remaining_tiles! This will avoid deleting duplicates of a letter present in remaining_tiles (with delete or - methods, for example) because we are getting rid of an item at a particular INDEX. I believe '.delete_at' will do the same thing as 'slice' in this case.

Then we can compare the size of the word to 7 - (size of remaining_tiles) ! This is because the player starts with 7 tiles at the beginning of every played word, and remaining_tiles is what's left over after letter removal. We know this was a valid word submission if the sizes are the same!

**Comment by daphnegold** [comment]
Latest commit updates file structure so that RSpec is working and tests are passing. We are ready to be merged.

File: lib/player.rb (hunk 5 above)
Comment (Hamled) [inline lib/player.rb:26]: Awesome explanation, thanks!

While this is entirely a matter of taste and your understanding of intended audience... lines like this that are fairly "bulky" / "chunky" / full of code with very particular behavior are excellent candidates for having a comment in your code.

In this particular case what I would likely comment the line with:

``` ruby
# This removes the first element in remaining_tiles that matches the letter l
```

in case anyone was unfamiliar with the `.index` or `.slice!` methods.

File: lib/scrabble_master.rb (hunk 1 above)
Comment (Hamled) [inline lib/scrabble_master.rb]: If you have time (I know we've move on already), I'd recommend double checking that win condition. I did manage to get to 100 points in my rspec run, but it kept asking me to play after that...
